src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractExtensionMethodImportCompletionProvider.cs

```diff
@@ -0,0 +1,277 @@
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal class ExtensionMethodImportCompletionService
+    {
+        // return a multi-dictionary of mappings "FQN of containing type" => "extension method names"
+        public async Task<MultiDictionary<string, string>> GetPossibleExtensionMethodMatchesAsync(
+            Project currentProject,
+            ISet<string> targetTypeNames,
+            CancellationToken cancellationToken)
+        {
+            var solution = currentProject.Solution;
+            var graph = currentProject.Solution.GetProjectDependencyGraph();
+            var dependencies = graph.GetProjectsThatThisProjectTransitivelyDependsOn(currentProject.Id);
+            var relevantProjects = dependencies.Select(solution.GetProject)
+                                                 .Where(p => p.SupportsCompilation)
+                                                 .Concat(currentProject);
+
+            var results = new MultiDictionary<string, string>();
+
+            // Find matching extension methods from source.
+            foreach (var project in relevantProjects)
+            {
+                foreach (var document in project.Documents)
+                {
+                    var info = await document.GetSyntaxTreeIndexAsync(cancellationToken).ConfigureAwait(false);
+                    if (!info.ContainsExtensionMethod)
+                    {
+                        continue;
+                    }
+
+                    // Add simple extension methods with matching target type name
+                    foreach (var targetTypeName in targetTypeNames)
+                    {
+                        if (!info.SimpleExtensionMethodInfo.TryGetValue(targetTypeName, out var methodInfoIndices))
+                        {
+                            continue;
+                        }
+
+                        foreach (var index in methodInfoIndices)
+                        {
+                            if (info.TryGetDeclaredSymbolInfo(index, out var methodInfo))
+                            {
+                                results.Add(methodInfo.FullyQualifiedContainerName, methodInfo.Name);
+                            }
+                        }
+                    }
+
+                    // Add all complex extension methods, we will need to completely rely on symbols to match them.
+                    foreach (var index in info.ComplexExtensionMethodInfo)
+                    {
+                        if (info.TryGetDeclaredSymbolInfo(index, out var methodInfo))
+                        {
+                            results.Add(methodInfo.FullyQualifiedContainerName, methodInfo.Name);
+                        }
+                    }
+                }
+            }
+
+            // Find matching extension methods from metadata
+            foreach (var peReference in currentProject.MetadataReferences.OfType<PortableExecutableReference>())
```

Comment:
what about metadata references from other projects that we aren't referencing in this project?

Reply:
You mean "metadata references from other projects (that are referenced by this project) that we aren't referencing in this project"? Are those not included in `currentProject.MetadataReferences`?

Reply:
Good to know, will need to add a test for this. Thanks!